controllers/nifinodegroupautoscaler_controller.go

```diff
@@ -0,0 +1,428 @@
+	err := r.Client.Get(ctx, req.NamespacedName, nodeGroupAutoscaler)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return Reconciled()
+		}
+		// Error reading the object - requeue the request.
+		return RequeueWithError(r.Log, err.Error(), err)
+	}
+
+	// Check if marked for deletion and run finalizers
+	if k8sutil.IsMarkedForDeletion(nodeGroupAutoscaler.ObjectMeta) {
+		return r.checkFinalizers(ctx, nodeGroupAutoscaler)
+	}
+
+	// Ensure finalizer for cleanup on deletion
+	if !util.StringSliceContains(nodeGroupAutoscaler.GetFinalizers(), autoscalerFinalizer) {
+		r.Log.Info(fmt.Sprintf("Adding Finalizer for NifiNodeGroupAutoscaler node group %s", nodeGroupAutoscaler.Spec.NodeConfigGroupId))
+		nodeGroupAutoscaler.SetFinalizers(append(nodeGroupAutoscaler.GetFinalizers(), autoscalerFinalizer))
+	}
+
+	// Get the last configuration viewed by the operator.
+	o, err := patch.DefaultAnnotator.GetOriginalConfiguration(nodeGroupAutoscaler)
+	// Create it if not exist.
+	if o == nil {
+		if err := patch.DefaultAnnotator.SetLastAppliedAnnotation(nodeGroupAutoscaler); err != nil {
+			return RequeueWithError(r.Log, "could not apply last state to annotation", err)
+		}
+		if err := r.Client.Update(ctx, nodeGroupAutoscaler); err != nil {
+			return RequeueWithError(r.Log, "failed to update NifiNodeGroupAutoscaler", err)
+		}
+		o, err = patch.DefaultAnnotator.GetOriginalConfiguration(nodeGroupAutoscaler)
+	}
+
+	// Check if the cluster reference changed.
+	original := &v1alpha1.NifiNodeGroupAutoscaler{}
+	json.Unmarshal(o, original)
+	if !v1alpha1.ClusterRefsEquals([]v1alpha1.ClusterReference{original.Spec.ClusterRef, nodeGroupAutoscaler.Spec.ClusterRef}) {
+		nodeGroupAutoscaler.Spec.ClusterRef = original.Spec.ClusterRef
+	}
+
+	// lookup NifiCluster reference
+	// we do not want cached objects here. We want an accurate state of what the cluster is right now, so bypass the client cache by using the APIReader directly.
+	cluster := &v1alpha1.NifiCluster{}
+	err = r.APIReader.Get(ctx,
+		types.NamespacedName{
+			Name:      nodeGroupAutoscaler.Spec.ClusterRef.Name,
+			Namespace: nodeGroupAutoscaler.Spec.ClusterRef.Namespace,
+		},
+		cluster)
+	if err != nil {
+		return RequeueWithError(r.Log, fmt.Sprintf("failed to look up cluster reference %v+", nodeGroupAutoscaler.Spec.ClusterRef), err)
+	}
+
+	// Handle HorizontalAutoScaler
+	hpa, err := r.horizontalPodAutoscaler(r.Log, cluster, nodeGroupAutoscaler)
+	if err != nil {
+		return RequeueWithError(r.Log, "failed to generate horizontal pod autoscaler spec", err)
+	}
+	err = k8sutil.Reconcile(r.Log, r.Client, hpa, nil)
+	if err != nil {
+		return RequeueWithError(r.Log, "failed to reconcile horizontal pod autoscaler", err)
+	}
```

Comment:
I think this is a good point. I was just following what seemed to be the approach for the `NifiCluster` since it creates things like `PodDisruptionBudget`.

I guess if we didn't create the HPA here, then users would have to create a `NifiNodeGroupAutoscaler` and a `HorizontalPodAutoscaler` with a `scaleTargerRef` that's pointed at that `NifiNodeGroupAutoscaler`. 

I suppose for the same reason you may not want to accept PR #84 to have the operator create a `ServiceMonitor` and rather leave that to a helm chart. Is that correct?

Reply:
Yes, exactly! 
(As soon as I find some time to do this, I'll add a helm chart in the repo that wrap the NiFiCluster and ServiceMonitor definition, which is why I haven't responded to your PR yet).

The `PodDisruptionBudget` is handled by the operator, as we directly handle the definition of a node that is a `Pod` and it is closely coupled.

Reply:
That makes sense. I will close #84.